MoleculeVAE-reImplementation/train.py

```diff
@@ -1,96 +1,330 @@
+    '''
+    :param xi,yi,zi: input data
+    :param c: property value
+    :param label: property name
+    :param title: id of epochs
+    :param n_min: min value of input property
+    :param n_max: max value of input property
+
+    To save animation to Gif:(ex.)
+    logp_frames=[]
+        #Training process start
+     logp_frames.append(
+            plot3D(plot_vector[:, 0], plot_vector[:, 1], plot_vector[:, 2], pre_plot[:, 0],
+                   "logP", str(epochs), -7, 9))
+        # End of the training process
+     gif.save(logp_frames, 'logp.gif', duration=150)
+
+    '''
+    data = pd.DataFrame(
+        {'x': xi, 'y': yi, 'z': zi, 'c': c})
+    fig = plt.figure()
+    ax = fig.add_subplot(projection='3d')
+    p = ax.scatter(data.x, data.y, data.z, c=data.c, alpha=0.6, s=10, cmap='YlGnBu', vmin=n_min, vmax=n_max)
+    ax.view_init(0, 45)
+    plt.colorbar(p, label=label)
+    # plt.axis('off')
+    plt.title('Epoch=' + title)
+
+
+'''========================End of loss module=========================='''
+
+'''Train & Test func.'''
+
+
+def train(epochs, train_set_final, valid_set_final):
     model.train()
     train_loss = 0
+    valid_loss = 0
     if params['do_prop_pred']:
-        X_train, X_test, Y_train, Y_test,od = Smiles2dataset(params)
-        text_labels_df = pd.DataFrame({'Smi': [X_train], 'PreLabel': [Y_train]})
-        test_labels_df = pd.DataFrame({'Smi': [X_test], 'PreLabel': [Y_test]})
-        dataset = CustomTextDataset(text_labels_df['Smi'], text_labels_df['PreLabel'])
-        testDataset = CustomTextDataset(test_labels_df['Smi'], test_labels_df['PreLabel'])
-        train_loader = torch.utils.data.DataLoader(dataset, batch_size=params['batch_size'])
-        test_loader = torch.utils.data.DataLoader(testDataset)
-        for batch_idx, (data,label) in enumerate(train_loader):
+        for batch_idx, (data, label) in enumerate(train_set_final):
             data = data.to(dtype=torch.float32, device=device)
-            label=torch.from_numpy(np.array(label)).to(device=device,dtype=torch.float32)
+            label = label.to(dtype=torch.float32, device=device)
             optimizer.zero_grad()
-            output, mean, logvar,pre = model(data)
-            pre_loss_ca=nn.MSELoss()
-            pre_loss=pre_loss_ca(pre,label)
-            loss = vae_loss(output.to(torch.float32), data, mean, logvar)+pre_loss
+            output, mean, logvar, pre, z = model(data)
+            pre_loss_ca = nn.MSELoss()
+            pre_loss = pre_loss_ca(pre, label)
+            loss = calculate_loss_by_type(output, data, mean, logvar, pre_loss, epochs, 'train')
             loss.backward()
             train_loss += loss
             optimizer.step()
-            if batch_idx % 1000 == 0:
-                torch.save(model.state_dict(), 'Weights/param.pth')
-                print(f'{epochs} / {batch_idx}\t{loss:.4f}')
+            if batch_idx % len(train_set_final) == 0:
```

Comment:
Putting inside this `if` is a bit odd, as it just means it will get triggered at the end of the loop. Why not put it in the epoch look but outside the train set loop?